ruby/ruby_assign.rb

```diff
@@ -1,265 +1,254 @@
-  maths: 43,
-  physics: 54,
-  chemistry:65,
-  year: 2016,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 2,
-  department: "a1",
-  maths: 66,
-  physics: 52,
-  chemistry:65,
-  year: 2016,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 3,
-  department: "a7",
-  maths: 87,
-  physics: 32,
-  chemistry:43,
-  year: 2016,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 1,
-  department: "a1",
-  maths: 21,
-  physics: 52,
-  chemistry:65,
-  year: 2015,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 2,
-  department: "a1",
-  maths: 68,
-  physics: 50,
-  chemistry:65,
-  year: 2015,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 3,
-  department: "a7",
-  maths: 85,
-  physics: 22,
-  chemistry:43,
-  year: 2015,
-  count: 1
-}
-records << record
-
-record = {
-  student_id: 4,
-  department: "a7",
-  maths: 21,
-  physics: 22,
-  chemistry:13,
-  year: 2016,
-  count: 1
-}
-records << record
-
-#starting of the program
+#taking all the required user input
+def group_by_input
+  begin
+    print "Enter valid Group_by ?  "
+    group_bi = gets.chomp
+    group_bi.downcase!
+  end until (["student_id","department","year"].include? group_bi)
+  return group_bi
+end
 
-#taking user input
-print "Group_by ?  "
-group_bi = gets.chomp
-group_bi.downcase!
+def sort_by_input
+  begin
+    print "Enter valid Sort_by ?  "
+    sort_bi = gets.chomp
+    sort_bi.downcase!
+  end until (["physics", "chemistry", "maths"].include? sort_bi)
+  return sort_bi
+end
 
-while (group_bi != "student_id" && group_bi != "department" && group_bi != "year")
-  print "Enter valid Group_by ?  "
-  group_bi = gets.chomp
-  group_bi.downcase!
+def display_fields_input(records)
+  begin
+    print "Enter all valid Display_fields ?  "
+    display = gets.chomp
+    display.downcase!
+    display_fields = display.split(",")
+    flag = true
+
+    display_fields.each do |field|
+       (puts "#{field} is not a valid field"
+        flag = false) unless records[0].has_key? field.to_sym
+    end
+  end until flag
+  return display_fields
 end
 
-print "Sort_by ?  "
-sort_bi = gets.chomp
-sort_bi.downcase!
+def should_compare_input
+  begin
+    print "Should_compare (true/false) ? "
+    should_compare = gets.chomp
+    should_compare.downcase!
+  end until (["true","false"].include? should_compare)
+  return (should_compare == "true"? true : false)
+end
 
-while (sort_bi != "maths" && sort_bi != "physics" && sort_bi != "chemistry")
-  print "Enter valid Group_by ?  "
-  sort_bi = gets.chomp
-  sort_bi.downcase!
+def compare_on_input
+  begin
+    print "Enter valid field to Compare_on ?  "
+    compare_on = gets.chomp
+    compare_on.downcase!
+  end until (["student_id","department","year"].include? compare_on)
+  return compare_on
 end
 
-print "Display_fields ?  "
-display = gets.chomp
-display.downcase!
-display_fields = display.split(",")
+def first_element_input(records, compare_on)
```

Comment:
reduce the functions first_element_input, second_element_input to single function